keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,508 @@
+which would ensure that the PVCs created automatically during the StatefulSet activation
+is deleted once the StatefulSet is deleted.
+
+#### Story 2
+User is very cost conscious and can sustain slower scale-up speeds, even after a
+scale-down. The user does not want to pay for volumes that are not in use, and
+so wants them to be reclaimed as soon as possible, including during
+scale-down. On scale-up a new volume will be provisioned and the new pod will
+have to re-intitialize. However, for short-lived interruptions when a pod is
+killed & recreated, like a rolling update or node disruptions, the data on
+volumes is persisted. This is a key property that ephemeral storage, like
+emptyDir, cannot provide.
+
+User would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnScaledown` ensuring
+PVCs are deleted when corresponding Pods are deleted. New Pods created during scale-up
+followed by a scale-down will wait for freshly created PVCs.
+
+### Notes/Constraints/Caveats (optional)
+
+This feature applies to PVCs which are dynamically provisioned from the
+volumeClaimTemplate of a StatefulSet. Any PVC and PV provisioned from this
+mechanism will function with this feature. Note that to current knowledge these
+PVCs are identified by the static naming scheme used by StatefulSets. If a user
+pre-provisions a PVC with the same name, it will be treated as an
+auto-provisioned one and used accordingly. It is not clear what happens today if
+a user has pre-provisioned a PVC matching one that the StatefulSet will try to
+create.
+
+### Risks and Mitigations
+
+Currently the PVCs created by StatefulSet are not deleted automatically. Using the
+`DeleteOnScaledown` or `DeleteOnStatefulSetDeletion` would delete the PVCs
+automatically. Since this involves persistent data being deleted, users should take
+appropriate care using this feature. Having the `Retain` behaviour as default will ensure
+that the PVCs remain intact by default and only a conscious choice made by user will
+involve any persistent data being deleted.
+
+This proposed API causes the PVCs associated with the StatefulSet to have
+behavior close to, but not the same as, ephemeral volumes, such as emptyDir or
+generic ephemeral volumes. This may cause user confusion. PVCs under this policy
+will more durable than ephemeral volumes would be, as they are only deleted on
+scale-down or StatefulSet deletion, and not on other pod lifecycle events like
+being rescheduled to a new node, even with the new retain policies.
+
+## Design Details
+
+### Volume delete policy for the StatefulSet created PVCs
+
+A new field named `PersistentVolumeClaimDeletePolicy` of the type
+`StatefulSetPersistentVolumeClaimDeletePolicy` will be added to the StatefulSet. This
+field will represent the user indication on whether the associated PVCs can be
+automatically deleted or not. The default policy would be `Retain`.
+
+In alpha release we intend to keep the `PersistentVolumeClaimDeletePolicy` immutable after
```

Comment:
I wonder if we should allow it to be mutable. We have PV reclaimPolicy as mutable because folks may originally have decided on Delete, but sometime later need to do some sort of manual repair/rollback but still keep the data, and want to temporarily switch to Retain.

If we make it immutable + reconciling, then it will be very difficult to do this. Of course, if we do make it mutable, there is still going to be a time lag for the "onDelete" policy because the controller has to go and remove the ownerRef on all the PVCs. Maybe that's another argument to only add the ownerRef on demand?

Reply:
As discussed offline, you raise a good point. Mutability is a good idea. Since we've also decided that the ownerRefs will be reconciled and not just made on PVC creation, this seems to not introduce too many complexities.